### PR TITLE
[bilibili.py]interaction video support

### PR DESCRIPTION
**Pull request by ljhcage**
Support interaction video for bilibili, BFS all possible node and download video. use --playlist option
支持B站的互动视频，使用广度优先搜索所有可能的节点，然后下载对应的视频，使用playlist指令即可

### CODE REVIEW

**Comment by blueset** [comment]
Despite this is not the fault of this PR, I'd still comment it here for the sake of simplicity.

This fails on `https://www.bilibili.com/video/av62990149` as some branches of the video contains **no audio**.

Probably easier to get rid of the `audio_baseurl` when adding the stream in the case where audio is not present. 

https://github.com/soimort/you-get/blob/07aa5b06841cb4ee29a0c4114d1f5680d1ad2026/src/you_get/extractors/bilibili.py#L258-L268

Traceback:

```
Traceback (most recent call last):
  File "/usr/local/bin/you-get", line 10, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.7/site-packages/you_get/__main__.py", line 92, in main
    main(**kwargs)
  File "/usr/local/lib/python3.7/site-packages/you_get/common.py", line 1758, in main
    script_main(any_download, any_download_playlist, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/you_get/common.py", line 1646, in script_main
    **extra
  File "/usr/local/lib/python3.7/site-packages/you_get/common.py", line 1300, in download_main
    download_playlist(url, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/you_get/common.py", line 1754, in any_download_playlist
    m.download_playlist(url, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/you_get/extractors/bilibili.py", line 590, in download_playlist_by_url
    self.prepare_by_cid(aid, initial_state['videoData']['cid'], initial_state['videoData']['title'] + ('P{}. {}'.format(1, node_info['data']['title'])),html_content,playinfo,playinfo_,url)
  File "/usr/local/lib/python3.7/site-packages/you_get/extractors/bilibili.py", line 495, in prepare_by_cid
    audio_baseurl = playinfo['data']['dash']['audio'][0]['baseUrl']
TypeError: 'NoneType' object is not subscriptable
```

**Comment by sontekliu** [comment]
I have a problem at download video on bilibili.



Who can help me? Thanks

**Comment by ljhcage** [comment]
> Despite this is not the fault of this PR, I'd still comment it here for the sake of simplicity.
> 
> This fails on `https://www.bilibili.com/video/av62990149` as some branches of the video contains **no audio**.
> 
> Probably easier to get rid of the `audio_baseurl` when adding the stream in the case where audio is not present.
> 
> https://github.com/soimort/you-get/blob/07aa5b06841cb4ee29a0c4114d1f5680d1ad2026/src/you_get/extractors/bilibili.py#L258-L268
> 
> Traceback:
> 
> ```
> Traceback (most recent call last):
>   File "/usr/local/bin/you-get", line 10, in <module>
>     sys.exit(main())
>   File "/usr/local/lib/python3.7/site-packages/you_get/__main__.py", line 92, in main
>     main(**kwargs)
>   File "/usr/local/lib/python3.7/site-packages/you_get/common.py", line 1758, in main
>     script_main(any_download, any_download_playlist, **kwargs)
>   File "/usr/local/lib/python3.7/site-packages/you_get/common.py", line 1646, in script_main
>     **extra
>   File "/usr/local/lib/python3.7/site-packages/you_get/common.py", line 1300, in download_main
>     download_playlist(url, **kwargs)
>   File "/usr/local/lib/python3.7/site-packages/you_get/common.py", line 1754, in any_download_playlist
>     m.download_playlist(url, **kwargs)
>   File "/usr/local/lib/python3.7/site-packages/you_get/extractors/bilibili.py", line 590, in download_playlist_by_url
>     self.prepare_by_cid(aid, initial_state['videoData']['cid'], initial_state['videoData']['title'] + ('P{}. {}'.format(1, node_info['data']['title'])),html_content,playinfo,playinfo_,url)
>   File "/usr/local/lib/python3.7/site-packages/you_get/extractors/bilibili.py", line 495, in prepare_by_cid
>     audio_baseurl = playinfo['data']['dash']['audio'][0]['baseUrl']
> TypeError: 'NoneType' object is not subscriptable
> ```

fixed. I'll pull a new request after this is merged

**Comment by ljhcage** [comment]
@sontekliu can't reproduce, works fine for me, try again please, might be network problem.
我这边是没有问题的，你再多试试看，可能是网络问题。

**Comment by soimort** [comment]
Thanks!

File: src/you_get/extractors/bilibili.py [hunk 1]
```diff
@@ -474,9 +570,66 @@ def download_playlist_by_url(self, url, **kwargs):
...
+                    'screen': 0,
+                }
+                node_info = json.loads(get_content('https://api.bilibili.com/x/stein/nodeinfo?'+parse.urlencode(params)))
+
+                playinfo_text = match1(html_content, r'__playinfo__=(.*?)</script><script>')  # FIXME
+                playinfo = json.loads(playinfo_text) if playinfo_text else None
+
+                html_content_ = get_content(self.url, headers=self.bilibili_headers(cookie='CURRENT_FNVAL=16'))
+                playinfo_text_ = match1(html_content_, r'__playinfo__=(.*?)</script><script>')  # FIXME
+                playinfo_ = json.loads(playinfo_text_) if playinfo_text_ else None
```
Comment (soimort) [inline src/you_get/extractors/bilibili.py:596]: Reusing `playinfo` and `playinfo_` causes a serious bug for regular multi-part videos: if the 2nd video does not have the same DASH stream quality as the 1st one, `you-get -l` will end up downloading the same video (1st one) for all parts.
 
Examples:
- you-get -l https://www.bilibili.com/video/BV1fc41187xA  
- you-get -l https://www.bilibili.com/video/BV1ms411J7V3

You get the same video for all parts.

This has been a longstanding issue I haven't noticed before; Interaction videos should be handled separately and not to be confused with regular multi-part videos.

**Pull request merged**